File: totoggle_node/src/context.ts
/** Request/deployment information used to evaluate contextual activation rules. */
export interface ToggleContext {
  readonly rolloutKey?: string;
  readonly parameter?: string;
  readonly userId?: string;
  readonly ip?: string;
  readonly country?: string;
  /** Deployment or request cohort, e.g. "canary", "beta", or "stable". */
  readonly cohort?: string;
  readonly attributes?: Readonly<Record<string, string>>;
}

/**
 * Application-owned bridge from its request/deployment context to ToToggle. SDKs deliberately
 * do not inspect HTTP headers, sessions, or proxy addresses themselves: those are framework- and
 * deployment-specific and trusting them blindly is unsafe.
 */
export interface ToggleContextProvider {
  getContext(): ToggleContext | undefined;
}

File: totoggle_go/context.go
package totoggle

import "context"

// ToggleContext is request/deployment information used only by the rule on the requested
// toggle. Applications populate it through ToggleContextProvider; the SDK never guesses HTTP
// headers, proxy addresses, or authentication state.
type ToggleContext struct {
	RolloutKey string
	Parameter  string
	UserID     string
	IP         string
	Country    string
	Cohort     string
	Attributes map[string]string
}

// ToggleContextProvider bridges an application's middleware/request context to ToToggle.
// Returning nil means no context is available and contextual rules fail closed.
type ToggleContextProvider interface {
	ToggleContext(context.Context) *ToggleContext
}
